api-fput-object.go

```diff
@@ -128,60 +86,178 @@ func (c Client) FPutObject(bucketName, objectName, filePath, contentType string)
 
 	// Save the file size.
 	fileSize := fileStat.Size()
-	var enableSha256Sum bool
-	if !c.signature.isV2() {
-		enableSha256Sum = true
+	if fileSize > int64(maxMultipartPutObjectSize) {
+		return 0, ErrInvalidArgument("Input file size is bigger than the supported maximum of 5TB.")
+	}
+
+	// NOTE: Google Cloud Storage multipart Put is not compatible with Amazon S3 APIs.
+	// Current implementation will only upload a maximum of 5GB to Google Cloud Storage servers.
+	if isGoogleEndpoint(c.endpointURL) {
+		if fileSize <= -1 || fileSize > int64(maxSinglePutObjectSize) {
+			return 0, ErrorResponse{
+				Code:       "NotImplemented",
+				Message:    fmt.Sprintf("Invalid Content-Length %d for file uploads to Google Cloud Storage.", fileSize),
+				Key:        objectName,
+				BucketName: bucketName,
+			}
+		}
+		// Do not compute MD5 for Google Cloud Storage. Uploads upto 5GB in size.
+		n, err := c.putNoChecksum(bucketName, objectName, fileData, fileSize, contentType)
+		return n, err
+	}
+
+	// NOTE: S3 doesn't allow anonymous multipart requests.
+	if isAmazonEndpoint(c.endpointURL) && c.anonymous {
+		if fileSize <= -1 || fileSize > int64(maxSinglePutObjectSize) {
+			return 0, ErrorResponse{
+				Code:       "NotImplemented",
+				Message:    fmt.Sprintf("For anonymous requests Content-Length cannot be %d.", fileSize),
+				Key:        objectName,
+				BucketName: bucketName,
+			}
+		}
+		// Do not compute MD5 for anonymous requests to Amazon S3. Uploads upto 5GB in size.
+		n, err := c.putAnonymous(bucketName, objectName, fileData, fileSize, contentType)
+		return n, err
+	}
+
+	// Large file upload is initiated for uploads for input data size
+	// if its greater than 5MB or data size is negative.
+	if fileSize >= minimumPartSize || fileSize < 0 {
+		n, err := c.fputLargeObject(bucketName, objectName, fileData, fileSize, contentType)
+		return n, err
+	}
+	n, err := c.putSmallObject(bucketName, objectName, fileData, fileSize, contentType)
+	return n, err
+}
+
+// sectionSums carries checksums for sectionReader.
+type sectionSums struct {
+	MD5Sum    []byte
+	Sha256Sum []byte
+	Size      int64
+}
+
+// sectionSumReader - calculates MD5 and Sha256 for an input section reader.
+func (c Client) sectionSumReader(sectionReader *io.SectionReader) (sectionSums, error) {
```

Comment:
sectionSumReader like naming convention is used interface names. Also let sectionreader be inside this function. Make this function as computeHash and return the values directly as variables. It is ok to return 4 variables here in this case. It will be more readable.